src/bijectors/coupling_layer.jl

```diff
@@ -0,0 +1,244 @@
+  [1, 1]  =  1.0,
+  [2, 1]  =  1.0,
+  [3, 1]  =  1.0)
+
+julia> # Partition into 3 parts; the last part is inferred to be indices `[3, ]` from
+       # the fact that `[1]` and `[2]` does not make up all indices in `1:3`.
+       x1, x2, x3 = partition(m, [1., 2., 3.])
+([1.0], [2.0], [3.0])
+
+julia> # Recombines the partitions into a vector
+       combine(m, x1, x2, x3)
+3-element Array{Float64,1}:
+ 1.0
+ 2.0
+ 3.0
+```
+
+"""
+struct PartitionMask{A}
+    A_1::A
+    A_2::A
+    A_3::A
+
+    # Only make it possible to construct using matrices
+    PartitionMask(A_1::A, A_2::A, A_3::A) where {A <: AbstractMatrix{<:Real}} = new{A}(A_1, A_2, A_3)
+end
+
+function PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::AbstractVector{Int}
+)
+    A_1 = spzeros(n, length(indices_1));
+    A_2 = spzeros(n, length(indices_2));
+    A_3 = spzeros(n, length(indices_3));
+
+    for (i, idx) in enumerate(indices_1)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_3)
+        A_3[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, A_3)
+end
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int}
+) = PartitionMask(n, indices_1, indices_2, nothing)
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::Nothing
+) = PartitionMask(n, indices_1, indices_2, [i for i in 1:n if i ∉ (indices_1 ∪ indices_2)])
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::Nothing,
+    indices_3::AbstractVector{Int}
+) = PartitionMask(n, indices_1, [i for i in 1:n if i ∉ (indices_1 ∪ indices_3)], indices_3)
+
+"""
+    PartitionMask(n::Int, indices)
+
+Assumes you want to _split_ the vector, where `indices` refer to the 
+parts of the vector you want to apply the bijector to.
+"""
+function PartitionMask(n::Int, indices)
+    indices_2 = [i for i in 1:n if i ∉ indices]
+
+    # sparse arrays <3
+    A_1 = spzeros(n, length(indices));
+    A_2 = spzeros(n, length(indices_2));
+
+    # Like doing:
+    #    A[1, 1] = 1.0
+    #    A[3, 2] = 1.0
+    for (i, idx) in enumerate(indices)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, spzeros(n, 0))
+end
+PartitionMask(x::AbstractVector, indices) = PartitionMask(length(x), indices)
+
+"""
+    combine(m::PartitionMask, x_1, x_2, x_3)
+
+Combines `x_1`, `x_2`, and `x_3` into a single vector.
+"""
+@inline combine(m::PartitionMask, x_1, x_2, x_3) = m.A_1 * x_1 .+ m.A_2 * x_2 .+ m.A_3 * x_3
+
+"""
+    partition(m::PartitionMask, x)
+
+Partitions `x` into 3 disjoint subvectors.
+"""
+@inline partition(m::PartitionMask, x) = (transpose(m.A_1) * x, transpose(m.A_2) * x, transpose(m.A_3) * x)
```

Comment:
This is not type-safe. If `x` is `Array{Float32}`, paritions are returned as `Array{Float64}`. Can you fix this?

Reply:
Isn't this correct behavior though? I see what you're saying, but think the fix is to just define the partitions also be `Float32`, no?

Reply:
Why would we expect to return an array of a different type after partitioning?

Reply:
In the sense that `PartitionMask` also has a parameter in the type, we leave type-promotion to the multiplication.

I see what you're saying though, but I don't think we're going to be able to do that in a consistent way. What if we have `PartitionMask{SparseArray}` and `x::SparseVector`, how do we ensure that we also return a `SparseVector`? Yeah, we can implement that, but we're not going to be able to cover all these different combinations, right?

Reply:
> Yeah, we can implement that, but we're not going to be able to cover all these different combinations, right?

I really mean the `eltype` should not be changed.

Reply:
If we only doing partitioning, it is expected.

Reply:
> I really mean the eltype should not be changed.
Yeah, but I honestly don't see how we can do that consistently :confused: 

And I don't think it's such a huge problem as this is something that generally comes up with transformations using matrix multiplication and the type is right there in the definition of `PartitionArray{A}`. It' similar to how if you do `Shift{Float64}  and use that on a `Float32`, things will be converted.

I _do_ however agree that with partitioning it's not so natural to think that this will make a difference, but, again, I don't see how we can consistently ensure that we don't change the `eltype` when we're using matrix multiplication under the hood :confused: 

To do that, I think we'll have to drop the use of matrix multiplication, but that also means that we drop GPU support, right? Could potentially define specific ones when using `CuArray`?

Reply:
> I do however agree that with partitioning it's not so natural to think that this will make a difference, but, again, I don't see how we can consistently ensure that we don't change the eltype when we're using matrix multiplication under the hood

I assume you should be fine if the element type of the mask is `Bool` - e.g., `3f0 * true isa Float32` and `3.0 * true isa Float64`.

Reply:
> Yeah, but I honestly don't see how we can do that consistently 😕

> I assume you should be fine if the element type of the mask is `Bool`

Yes that's why I suggest changing mask to `Bool` in other comments.

Reply:
> I assume you should be fine if the element type of the mask is Bool - e.g., 3f0 * true isa Float32 and 3.0 * true isa Float64.

Ah, now I understand why you provided the other suggested change @xukai92; sorry about that! For a simple example I did see a slight perf difference, but probably not enough for us to not make the change. Do you know if it will work nicely on GPU?

Reply:
Yes it works with GPU but you would need `A_1`, `A_2` and `A_3` also on GPU, e.g. by doing so
```julia
import Flux
Flux.@functor PartitionMask
mask = PartitionMask(...) |> Flux.gpu
```
Of course if we make the whole coupling layer a functor it would automatically happen when do `Flux.gpu` on the whole layer.

Reply:
Ah, but what you suggest won't result in a sparse matrix :confused: I just blew the memory of my GPU out trying the example of GPU usage I have in another comment from above. And converting directly do sparse arrays on GPU doesn't seem to work:
```julia
julia> A_1
100×10 SparseMatrixCSC{Bool,Int64} with 10 stored entries:
  [1 ,  1]  =  1
  [2 ,  2]  =  1
  [3 ,  3]  =  1
  [4 ,  4]  =  1
  [5 ,  5]  =  1
  [6 ,  6]  =  1
  [7 ,  7]  =  1
  [8 ,  8]  =  1
  [9 ,  9]  =  1
  [10, 10]  =  1

julia> CUSPARSE.CuSparseMatrixCSC(A_1)
ERROR: MethodError: no method matching CuArrays.CUSPARSE.CuSparseMatrixCSC(::Array{Int64,1}, ::Array{Int64,1}, ::Array{Bool,1}, ::Tuple{Int64,Int64})
Closest candidates are:
  CuArrays.CUSPARSE.CuSparseMatrixCSC(::Array{Ti,1}, ::Array{Ti,1}, ::Array{T,1}, ::Tuple{Int64,Int64}) where {T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64}, Ti<:Integer} at /home/tor/.julia/packages/CuArrays/1njKF/src/sparse/array.jl:207
  CuArrays.CUSPARSE.CuSparseMatrixCSC(::CuArray{Ti,N,P} where P where N, ::CuArray{Ti,N,P} where P where N, ::CuArray{T,N,P} where P where N, ::Tuple{Int64,Int64}) where {T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64}, Ti<:Integer} at /home/tor/.julia/packages/CuArrays/1njKF/src/sparse/array.jl:208
  CuArrays.CUSPARSE.CuSparseMatrixCSC(::CuArray{Ti,N,P} where P where N, ::CuArray{Ti,N,P} where P where N, ::CuArray{T,N,P} where P where N, ::Any, ::Tuple{Int64,Int64}) where {T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64}, Ti<:Integer} at /home/tor/.julia/packages/CuArrays/1njKF/src/sparse/array.jl:209
Stacktrace:
 [1] CuArrays.CUSPARSE.CuSparseMatrixCSC(::SparseMatrixCSC{Bool,Int64}) at /home/tor/.julia/packages/CuArrays/1njKF/src/sparse/array.jl:219
 [2] top-level scope at REPL[56]:1
```

I guess we could be using `Bool` if on CPU and `Float32` on GPU? I remember there being an error where if the input was `CuArray{Float64}` and the partition masks were `CuArray{Float32}`, it would just fail silently, i.e. returned `nothing`. So I guess in the GPU case we're going to require the input and the partition mask to have the _same_  `eltype` anyways :confused:

Reply:
What do you think? Should I make it `Bool` for now and then we can deal with the pains of GPU later? In the mean-time, it's of course possible to provide the matrices directly and so it's _possible_ to use GPU with sparse-arrays, but slightly inconvenient.

Reply:
Let's make it `Bool` for now and to resolve the bad memory usage due to dense matrix issue on GPU later.

Reply:
Aight!